modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/schema/CacheIndexImpl.java

```diff
@@ -153,9 +158,43 @@ public Index queryIndex() {
 
             InlineIndex iidx = idx.unwrap(InlineIndex.class);
 
+            BPlusTree.TreeRowClosure<IndexRow, IndexRow> rowFilter = null;
+
+            if (notNull) {
+                boolean nullsFirst = collation.getFieldCollations().get(0).nullDirection ==
+                    RelFieldCollation.NullDirection.FIRST;
+
+                BPlusTree.TreeRowClosure<IndexRow, IndexRow> notNullRowFilter = IndexScan.createNotNullRowFilter(iidx);
+
+                AtomicBoolean skipCheck = new AtomicBoolean();
+
+                rowFilter = new BPlusTree.TreeRowClosure<IndexRow, IndexRow>() {
+                    @Override public boolean apply(
+                        BPlusTree<IndexRow, IndexRow> tree,
+                        BPlusIO<IndexRow> io,
+                        long pageAddr,
+                        int idx
+                    ) throws IgniteCheckedException {
+                        // If we have NULLS-FIRST collation, all values after first not-null value will be not-null,
+                        // don't need to check it with notNullRowFilter.
```

Comment:
Is there a specific test for this case with 'COUNT()'?

Reply:
Actually it tested in testCountIndexedField (some fields have NULL value), but I've added yet another test with NULL and two indexed columns.